src/debugger/crash_screen.c

```diff
@@ -0,0 +1,808 @@
+
+char *gCauseDesc[18] = {
+    "Interrupt",
+    "TLB modification",
+    "TLB exception on load",
+    "TLB exception on store",
+    "Address error on load",
+    "Address error on store",
+    "Bus error on inst.",
+    "Bus error on data",
+    "Failed Assert: See Assert Page",
+    "Breakpoint exception",
+    "Reserved instruction",
+    "Coprocessor unusable",
+    "Arithmetic overflow",
+    "Trap exception",
+    "Virtual coherency on inst.",
+    "Floating point exception",
+    "Watchpoint exception",
+    "Virtual coherency on data",
+};
+
+char *gFpcsrDesc[6] = {
+    "Unimplemented operation", "Invalid operation", "Division by zero", "Overflow", "Underflow",
+    "Inexact operation",
+};
+
+static u32 sProgramPosition = 0;
+static u32 sCrashScreenStackTraceCount = 0;
+
+static u16 gCrashScreenTextColor = 0xFFFF;
+static u32 sCrashScreenPrintRow_Pixels = 0;
+static u32 sCrashScreenPrintLnHeight_Pixels = GLYPH_HEIGHT;
+static struct {
+    OSThread thread;
+    u64 stack[THREAD2_STACK / sizeof(u64)];
+    OSMesgQueue mesgQueue;
+    OSMesg mesg;
+    u16 *framebuffer;
+    u16 width;
+    u16 height;
+} gCrashScreen;
+
+
+
+/**
+ * Splits a path string by the containing folder and the name of the file itself.
+ */
+char *crash_screen_ellide_string(char *str, u32 truncateLength) {
+    u32 string_length = strlen(str);
+
+    if (truncateLength >= string_length) {
+        return str;
+    }
+
+    str[string_length - truncateLength - 4] = '(';
+    str[string_length - truncateLength - 3] = '.';
+    str[string_length - truncateLength - 2] = '.';
+    str[string_length - truncateLength - 1] = '.';
+    str[string_length - truncateLength - 0] = ')';
+    return &str[string_length - truncateLength - 4];
+}
+
+static void set_text_color(u32 r, u32 g, u32 b) {
+    gCrashScreenTextColor = GPACK_RGBA5551(r, g, b, 255);
+}
+
+static void reset_text_color(void) {
+    gCrashScreenTextColor = 0xFFFF;
+}
+
+/**
+ * Sets the screen row, in pixels, at which the next print
+ *  will be placed on the Y axis.
+ */
+static void crash_screen_set_print_top(u32 row) {
+    sCrashScreenPrintRow_Pixels = row;
+}
+
+static void crash_screen_set_println_height(u32 height_pixels) {
+    sCrashScreenPrintLnHeight_Pixels = height_pixels;
+}
+
+static void crash_screen_reset_println_height(void) {
+    sCrashScreenPrintLnHeight_Pixels = GLYPH_HEIGHT;
+}
+
+void crash_screen_draw_rect(s32 x, s32 y, s32 w, s32 h) {
+    u16 *ptr;
+    s32 i, j;
+
+    ptr = gCrashScreen.framebuffer + gCrashScreen.width * y + x;
+    for (i = 0; i < h; i++) {
+        for (j = 0; j < w; j++) {
+            *ptr = 0x0001;
+            ptr++;
+        }
+        ptr += gCrashScreen.width - w;
+    }
+}
+
+void crash_screen_draw_glyph(s32 x, s32 y, s32 glyph) {
+    const u32 *data;
+    u16 *ptr;
+    u32 bit;
+    u32 rowMask;
+    s32 i, j;
+
+    if (glyph > 0x7F) return;
+
+    data = &sCrashScreenFont[((glyph&0xF)*GLYPH_HEIGHT * 2) + (glyph >= 64)];
```

Comment:
nit: I don't like how this reads; I think any conditionals used as arithmetic should _at the very least_ use a full ternary.